primitives/src/witness.rs

```diff
@@ -580,6 +576,34 @@ impl<'a> Arbitrary<'a> for Witness {
     }
 }
 
+/// Cast a decoded length prefix to a `usize`.
+///
+/// This function is basically just defensive. For all sane use cases the length prefix should be
+/// less than `MAX_VEC_SIZE` (on a 32-bit machine). If the value is bigger that `u16::MAX` and we
+/// are on a 16-bit machine you'll likely hit an error later anyway, better to just check it now.
+///
+/// # 16-bits
+///
+/// The compact size may be bigger than what can be represented in a `usize` on a 16-bit machine but
+/// this shouldn't happen if we created the witness becauese one would get an OOM error before that.
+// _I think_ the only way this could return `None` is if the witness is invalid anyways (i.e. length
+// prefix is for some reason not correct for the vector it represents).
+fn cast_to_usize_if_valid(n: u64) -> Option<usize> {
+    /// Maximum size, in bytes, of a vector we are allowed to decode.
+    const MAX_VEC_SIZE: u64 = 4_000_000;
+
+    if n > MAX_VEC_SIZE {
+        return None;
+    }
+
+    // 4,000,000 needs more than 16 bits.
+    if mem::size_of::<usize>() <= 2 {
+        None
```

Comment:
In bf6e959d777d3f51360d608ca454a021dfe3b792:

lmao then here we say 16-bit users aren't allowed to index witnesses *at all*?

Reply:
The point is that this returns `None` no matter what the original value was.

Reply:
Oh lol, I see it now. That's a fail.